flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/RedirectHandler.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.flink.shaded.netty4.io.netty.channel.SimpleChannelInboundHandler;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponse;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.router.KeepAliveWrite;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.router.Routed;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link SimpleChannelInboundHandler} which encapsulates the redirection logic for the
+ * REST endpoints.
+ *
+ * @param <T> type of the leader to retrieve
+ */
+@ChannelHandler.Sharable
+public abstract class RedirectHandler<T extends RestfulGateway> extends SimpleChannelInboundHandler<Routed> {
+
+	protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+	private final CompletableFuture<String> localAddressFuture;
+
+	protected final GatewayRetriever<T> leaderRetriever;
+
+	protected final Time timeout;
+
+	/** Whether the web service has https enabled. */
+	protected final boolean httpsEnabled;
+
+	private String localAddress;
+
+	protected RedirectHandler(
+			@Nonnull CompletableFuture<String> localAddressFuture,
+			@Nonnull GatewayRetriever<T> leaderRetriever,
+			@Nonnull Time timeout,
+			boolean httpsEnabled) {
+		this.localAddressFuture = Preconditions.checkNotNull(localAddressFuture);
+		this.leaderRetriever = Preconditions.checkNotNull(leaderRetriever);
+		this.timeout = Preconditions.checkNotNull(timeout);
+		this.httpsEnabled = httpsEnabled;
+		localAddress = null;
+	}
+
+	@Override
+	protected void channelRead0(
+		ChannelHandlerContext channelHandlerContext,
+		Routed routed) throws Exception {
+
+		try {
+			if (localAddressFuture.isDone()) {
+				if (localAddress == null) {
+					try {
+						localAddress = localAddressFuture.get(timeout.toMilliseconds(), TimeUnit.MILLISECONDS);
+					} catch (Exception e) {
+						logger.error("Could not obtain local address.", e);
+
+						HandlerUtils.sendErrorResponse(
+							channelHandlerContext,
+							routed.request(),
+							new ErrorResponseBody(new FlinkException("Could not obtain local address.", e)),
+							HttpResponseStatus.INTERNAL_SERVER_ERROR);
+					}
+				}
+
+				OptionalConsumer<T> optLeaderConsumer = OptionalConsumer.of(leaderRetriever.getNow());
+
+				optLeaderConsumer.ifPresent(
+					(T gateway) -> {
+						OptionalConsumer<CompletableFuture<String>> optRedirectAddressConsumer = OptionalConsumer.of(
+							HandlerRedirectUtils.getRedirectAddress(
+								localAddress,
+								gateway,
+								timeout));
+
+						optRedirectAddressConsumer
+							.ifPresent(
+								(CompletableFuture<String> redirectAddressFuture) ->
+									redirectAddressFuture.whenComplete(
```

Comment:
The HTTP redirection seems unnecessary. Why not pass the gateway to the handler in all cases? this would make the handlers act as a proxy to the actual leader which, as I understand, is the behavior we want long-term anyway.

Reply:
This would be possible if all REST handler related RPCs would return serializable values. Currently, some of the REST handler expect an `ExecutionGraph` as result and thus only work if the rest endpoint runs in the same `ActorSystem` as the `JobManager`. Once this has been changed, we no longer need the HTTP redirect. But right now, this is not the case.